near-sdk/src/store/tree_map/mod.rs

```diff
@@ -0,0 +1,2059 @@
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use near_sdk::store::TreeMap;
+    ///
+    /// let mut map: TreeMap<i32, i32> = TreeMap::new(b"t");
+    /// map.extend([4, 6, 8, 11]
+    ///     .iter()
+    ///     .map(|&s| (s, 0)));
+    /// for (_, balance) in map.range_mut(6..10) {
+    ///     *balance += 100;
+    /// }
+    /// for (id, balance) in &map {
+    ///     println!("{} => {}", id, balance);
+    /// }
+    /// ```
+    pub fn range_mut<R, Q>(&mut self, range: R) -> RangeMut<'_, K, V, H>
+    where
+        K: BorshDeserialize + Borrow<Q>,
+        Q: ?Sized + Ord,
+        R: RangeBounds<Q>,
+    {
+        RangeMut::new(self, (range.start_bound(), range.end_bound()))
+    }
+}
+
+impl<K, V, H> TreeMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    /// Removes a key from the map, returning the stored key and value if the
+    /// key was previously in the map.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match
+    /// those for the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use near_sdk::store::TreeMap;
+    ///
+    /// let mut map = TreeMap::new(b"m");
+    /// map.insert(1, "a".to_string());
+    /// assert_eq!(map.remove(&1), Some("a".to_string()));
+    /// assert_eq!(map.remove(&1), None);
+    /// ```
+    pub fn remove_entry<Q: ?Sized>(&mut self, key: &Q) -> Option<(K, V)>
+    where
+        K: Borrow<Q> + BorshDeserialize + Clone,
+        Q: BorshSerialize + ToOwned<Owned = K> + Eq + PartialOrd,
+    {
+        self.values.remove(key).map(|removed_value| {
+            let (new_root, removed) = self.tree.do_remove(key);
+            self.tree.root = new_root;
+            (expect(removed), removed_value)
+        })
+    }
+
+    /// Gets the given key's corresponding entry in the map for in-place manipulation.
+    /// ```
+    /// use near_sdk::store::TreeMap;
+    ///
+    /// let mut count = TreeMap::new(b"m");
+    ///
+    /// for ch in [7, 2, 4, 7, 4, 1, 7] {
+    ///     let counter = count.entry(ch).or_insert(0);
+    ///     *counter += 1;
+    /// }
+    ///
+    /// assert_eq!(count[&4], 2);
+    /// assert_eq!(count[&7], 3);
+    /// assert_eq!(count[&1], 1);
+    /// assert_eq!(count.get(&8), None);
+    /// ```
+    pub fn entry(&mut self, key: K) -> Entry<K, V>
+    where
+        K: Clone,
+    {
+        Entry::new(self.values.entry(key), &mut self.tree)
+    }
+}
+
+impl<K, V, H> TreeMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    /// Flushes the intermediate values of the map before this is called when the structure is
+    /// [`Drop`]ed. This will write all modified values to storage but keep all cached values
```

Comment:
On this topic, might be worth adding our usual `artibtrary`-based test suite here as well? I don't think `flush` is being tested right now.

Reply:
Drop wasn't added because everything in this function is implicitly called when dropped, so it is redundant. For maintainability reasons, I think I will just add the drop impl anyway to make sure if anything changes within `flush` that this is called.

As for arbitrary tests, these are basically covered by the quickcheck tests, and there would be a lot of overlap. Nonetheless will add in some